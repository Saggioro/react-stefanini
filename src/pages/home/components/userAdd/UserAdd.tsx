import { Formik } from "formik";
import React, { FunctionComponent } from "react";
import * as Yup from "yup";
import Button from "../../../../components/button/Button";
import Input from "../../../../components/input/Input";
import { useAppSelector } from "../../../../hooks/LocalReduxThunk";

import { Container } from "./UserAdd.styled";
import * as userActions from "../../../../store/actions/users";
import { useDispatch } from "react-redux";
import User from "../../../../models/User";
interface IUserAdd {}

const UserAdd: FunctionComponent<IUserAdd> = ({ ...props }: IUserAdd) => {
  const users = useAppSelector((state) => state.usersReducer.users);
  const dispatch = useDispatch();
  const initialState = { login: "", password: "", confirmPassword: "" };
  const validation = Yup.object().shape({
    login: Yup.string().required(),
    password: Yup.string().required(),
    confirmPassword: Yup.string().required(),
  });
  return (
    <Container>
      <Formik
        initialValues={initialState}
        onSubmit={async (values, { resetForm, setErrors }) => {
          try {
            const sameUser = users.find((user) => user.name === values.login);
            if (sameUser) {
              setErrors({ login: "This name is already being used" });
              throw new Error("Same user");
            }
            if (values.password !== values.confirmPassword) {
              setErrors({
                password: "Passwords must match",
                confirmPassword: "Passwords must match",
              });
              throw new Error("Password doesnt match");
            }
            dispatch(
              userActions.setUsers([
                ...users,
                new User(values.login, values.password),
              ]) as any
            );
            resetForm();
          } catch (e: any) {}
        }}
        validationSchema={validation}
      >
        {({ values, errors, touched, handleChange, handleSubmit }) => (
          <form onSubmit={handleSubmit}>
            <Input
              label="Login"
              value={values.login}
              name="login"
              onChange={handleChange}
              error={touched.login ? errors.login : undefined}
            />
            <Input
              label="Password"
              type="password"
              value={values.password}
              name="password"
              onChange={handleChange}
              error={touched.password ? errors.password : undefined}
            />
            <Input
              label="Confirm password"
              type="password"
              value={values.confirmPassword}
              name="confirmPassword"
              onChange={handleChange}
              error={
                touched.confirmPassword ? errors.confirmPassword : undefined
              }
            />
            <Button>Save</Button>
          </form>
        )}
      </Formik>
    </Container>
  );
};

export default UserAdd;
