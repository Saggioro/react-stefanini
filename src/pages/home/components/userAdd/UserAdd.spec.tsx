/* eslint-disable testing-library/no-unnecessary-act */
/* eslint-disable testing-library/no-node-access */
/* eslint-disable testing-library/no-container */
/* eslint-disable testing-library/prefer-screen-queries */
/* eslint-disable testing-library/no-debugging-utils */
import renderer, { act } from "react-test-renderer";
import { render, screen, fireEvent } from "@testing-library/react";
import UserAdd from "./UserAdd";

const mockDispatch = jest.fn();
jest.mock("react-redux", () => ({
  useSelector: jest.fn(),
  useDispatch: () => mockDispatch,
}));

jest.mock("../../../../hooks/LocalReduxThunk", () => {
  return {
    useAppSelector() {
      return [
        {
          name: "teste",
          password: "123",
        },
      ];
    },
  };
});
test("UserAdd should render correctly", () => {
  const component = renderer.create(<UserAdd />);

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("UserAdd form should reset after adding", async () => {
  const { container } = render(<UserAdd />);
  const mockedDispatch = jest.fn();
  mockDispatch.mockReturnValue(mockedDispatch);
  const inputLogin = container.querySelector(`input[name="login"]`) as Element;
  const button = container.querySelector(`button[type="submit"]`) as Element;
  const inputPassword = container.querySelector(
    `input[name="password"]`
  ) as Element;
  const inputConfirm = container.querySelector(
    `input[name="confirmPassword"]`
  ) as Element;
  await act(async () => {
    fireEvent.change(inputLogin, { target: { value: "admin" } });
    fireEvent.change(inputPassword, { target: { value: "123" } });
    fireEvent.change(inputConfirm, { target: { value: "123" } });
    fireEvent.click(button);
  });

  expect(inputLogin.value).toBe("");
  expect(screen.queryByText("123")).not.toBeInTheDocument();
});

test("UserAdd form should throw error when name exists", async () => {
  const { container } = render(<UserAdd />);
  const inputLogin = container.querySelector(`input[name="login"]`) as Element;
  const button = container.querySelector(`button[type="submit"]`) as Element;
  const inputPassword = container.querySelector(
    `input[name="password"]`
  ) as Element;
  const inputConfirm = container.querySelector(
    `input[name="confirmPassword"]`
  ) as Element;
  await act(async () => {
    fireEvent.change(inputLogin, { target: { value: "teste" } });
    fireEvent.change(inputPassword, { target: { value: "123" } });
    fireEvent.change(inputConfirm, { target: { value: "123" } });
    fireEvent.click(button);
  });

  expect(
    screen.getByText("This name is already being used")
  ).toBeInTheDocument();
});

test("UserAdd form should throw error when passwords doesnt match", async () => {
  const { container } = render(<UserAdd />);
  const inputLogin = container.querySelector(`input[name="login"]`) as Element;
  const button = container.querySelector(`button[type="submit"]`) as Element;
  const inputPassword = container.querySelector(
    `input[name="password"]`
  ) as Element;
  const inputConfirm = container.querySelector(
    `input[name="confirmPassword"]`
  ) as Element;
  await act(async () => {
    fireEvent.change(inputLogin, { target: { value: "teste2" } });
    fireEvent.change(inputPassword, { target: { value: "123" } });
    fireEvent.change(inputConfirm, { target: { value: "321" } });
    fireEvent.click(button);
  });

  expect(await screen.findAllByText("Passwords must match")).toHaveLength(2);
});
