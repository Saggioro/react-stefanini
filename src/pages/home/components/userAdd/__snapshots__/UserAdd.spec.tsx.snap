// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserAdd should render correctly 1`] = `
<div
  className="sc-jSMfEi cjUZxb"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="sc-gsnTZi crCdMa"
    >
      <label
        className="sc-hKMtZM hlXAcN"
      >
        Login
      </label>
      <input
        className="sc-dkzDqf bHWPMi"
        name="login"
        onChange={[Function]}
        type="text"
        value=""
      />
      <span
        className="sc-eCYdqJ gubPie"
      />
    </div>
    <div
      className="sc-gsnTZi crCdMa"
    >
      <label
        className="sc-hKMtZM hlXAcN"
      >
        Password
      </label>
      <input
        className="sc-dkzDqf bHWPMi"
        name="password"
        onChange={[Function]}
        type="password"
        value=""
      />
      <span
        className="sc-eCYdqJ gubPie"
      />
    </div>
    <div
      className="sc-gsnTZi crCdMa"
    >
      <label
        className="sc-hKMtZM hlXAcN"
      >
        Confirm password
      </label>
      <input
        className="sc-dkzDqf bHWPMi"
        name="confirmPassword"
        onChange={[Function]}
        type="password"
        value=""
      />
      <span
        className="sc-eCYdqJ gubPie"
      />
    </div>
    <button
      className="sc-bczRLJ eVNmrq"
      type="submit"
    >
      Save
    </button>
  </form>
</div>
`;
